File: 2025/src/day_10.py
from concurrent.futures import ProcessPoolExecutor
from dataclasses import dataclass
from pathlib import Path

from utils import benchmark


@dataclass(slots=True)
class Machine:
    lights: str
    buttons: list[tuple[int, ...]]
    joltage: tuple[int, ...]


@dataclass(slots=True)
class Data:
    machines: list[Machine]


@benchmark
def parse() -> Data:
    with open(Path(__file__).parent.parent / "inputs/day_10.txt") as f:
        file = f.read().strip().splitlines()
        machines: list[Machine] = []

        for line in file:
            lgt, *b, j = line.split()
            buttons = [tuple(map(int, button[1:-1].split(","))) for button in b]
            joltage = tuple(map(int, j[1:-1].split(",")))
            machines.append(Machine(lgt[1:-1], buttons, joltage))

    return Data(machines)


def push_button(lights: str, buttons: tuple[int, ...]) -> str:
    lgt = [lgt for lgt in lights]
    for b in buttons:
        lgt[b] = "#" if lights[b] == "." else "."

    return "".join(lgt)


def push_button_jolt(
    joltage: tuple[int, ...], buttons: tuple[int, ...]
) -> tuple[int, ...]:
    jlt = [j for j in joltage]
    for b in buttons:
        jlt[b] += 1

    return tuple(jlt)


def calc(machine: Machine):
    goal = machine.joltage
    buttons = machine.buttons
    button_presses = 0
    curr_joltage: set[tuple[int, ...]] = set([tuple([0 for _ in goal])])

    while True:
        curr_joltage = {
            push_button_jolt(jlt, btn) for jlt in curr_joltage for btn in buttons
        }

        button_presses += 1

        if goal in curr_joltage:
            print(f"completed {machine}")
            return button_presses


@benchmark
def part_1(data: Data) -> None:
    result = 0

    for machine in data.machines:
        button_presses = 0
        curr_lights: set[str] = set(["".join(["." for _ in machine.lights])])

        while True:
            curr_lights = {
                push_button(lgt, btn) for lgt in curr_lights for btn in machine.buttons
            }

            button_presses += 1

            if machine.lights in curr_lights:
                break

        result += button_presses

    print(f"Day 10, Part 1: {result}")


@benchmark
def part_2(data: Data) -> None:
    with ProcessPoolExecutor() as p:
        results = p.map(calc, data.machines)

    result = sum(results)
    print(f"Day 10, Part 2: {result}")


if __name__ == "__main__":
    data = parse()
    part_1(data)
    part_2(data)

File: 2025/src/day_10.rs
use std::{collections::HashSet, fs, time::Instant, vec};

use rayon::iter::{IntoParallelRefIterator, ParallelIterator};

fn part_1(points: &[Machine]) {
    let result = 0;

    println!("Day 10, Part 1: {}", result);
}

fn push_button_jolt(joltage: &[u32], buttons: &[u32]) -> Vec<u32> {
    let mut jlt = joltage.to_owned();
    for b in buttons {
        jlt[*b as usize] += 1;
    }

    jlt
}

fn calc(machine: &Machine) -> u32 {
    let mut button_presses = 0;
    let mut curr_joltage: HashSet<Vec<u32>> = HashSet::new();
    curr_joltage.insert(vec![0 as u32; machine.joltage.len()]);

    loop {
        let mut joltages: HashSet<Vec<u32>> = HashSet::new();
        for jlt in &curr_joltage {
            for btn in &machine.buttons {
                let new = push_button_jolt(jlt, btn);
                if new.iter().zip(&machine.joltage).any(|(a, b)| a > b) {
                    continue;
                }
                joltages.insert(new);
            }
        }

        curr_joltage = joltages;

        button_presses += 1;

        if curr_joltage.contains(&machine.joltage) {
            println!("completed {machine:?}");
            return button_presses;
        }
    }
}

fn part_2(machines: &[Machine]) {
    let result: u32 = machines.par_iter().map(|m| calc(m)).sum();

    println!("Day 10, Part 2: {}", result);
}

#[derive(Debug)]
struct Machine {
    buttons: Vec<Vec<u32>>,
    joltage: Vec<u32>,
}

impl Machine {
    fn new(buttons: Vec<Vec<u32>>, joltage: Vec<u32>) -> Self {
        Self { buttons, joltage }
    }
}

fn main() {
    let start = Instant::now();

    let file = fs::read_to_string("inputs/day_10.txt").expect("error reading file");
    let machines: Vec<Machine> = file
        .lines()
        .map(|l| {
            let split: Vec<&str> = l.split_whitespace().collect();

            let last = split.last().unwrap();
            let joltage: Vec<u32> = last[1..last.len() - 1]
                .split(",")
                .map(|n| n.parse::<u32>().unwrap())
                .collect();

            let buttons: Vec<Vec<u32>> = split[1..split.len() - 1]
                .into_iter()
                .map(|b| {
                    let buttons = b[1..b.len() - 1]
                        .split(",")
                        .map(|n| n.parse::<u32>().unwrap())
                        .collect();

                    buttons
                })
                .collect();

            Machine::new(buttons, joltage)
        })
        .collect();

    //part_1(&machines);
    part_2(&machines);

    println!("Finished in {}s", start.elapsed().as_secs());
}
